src/Stratis.Bitcoin.Features.Consensus/Rules/ConsensusRules.cs

```diff
@@ -0,0 +1,187 @@
+        /// The rules that will be registered with the rules engine.
+        /// </summary>
+        /// <returns>A list of rules.</returns>
+        IEnumerable<ConsensusRule> GetRules();
+    }
+
+    /// <inheritdoc />
+    public class ConsensusRules : IConsensusRules
+    {
+        /// <summary>A factory to creates logger instances for each rule.</summary>
+        private readonly ILoggerFactory loggerFactory;
+
+        /// <summary>Instance logger.</summary>
+        private readonly ILogger logger;
+
+        /// <summary>A collection of rules that well be executed by the rules engine.</summary>
+        private readonly Dictionary<ConsensusRule, ConsensusRule> consensusRules;
+
+        /// <summary>Specification of the network the node runs on - regtest/testnet/mainnet.</summary>
+        public Network Network { get; }
+
+        /// <summary>A provider of date and time.</summary>
+        public IDateTimeProvider DateTimeProvider { get; }
+
+        /// <summary>A chain of the longest block headers all the way to genesis.</summary>
+        public ConcurrentChain Chain { get; }
+
+        /// <summary>A deployment construction that tracks activation of features on the chain.</summary>
+        public NodeDeployments NodeDeployments { get; }
+
+        /// <summary>A collection of consensus constants.</summary>
+        public NBitcoin.Consensus ConsensusParams { get; }
+
+        /// <summary>The main loop of the consensus execution engine.</summary>
+        public ConsensusLoop ConsensusLoop { get; set; }
+
+        /// <summary>
+        /// Initializes an instance of the object.
+        /// </summary>
+        public ConsensusRules(Network network, ILoggerFactory loggerFactory, IDateTimeProvider dateTimeProvider, ConcurrentChain chain, NodeDeployments nodeDeployments)
+        {
+            this.Network = network;
+            this.DateTimeProvider = dateTimeProvider;
+            this.Chain = chain;
+            this.NodeDeployments = nodeDeployments;
+            this.loggerFactory = loggerFactory;
+            this.ConsensusParams = this.Network.Consensus;
+
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+            this.consensusRules = new Dictionary<ConsensusRule, ConsensusRule>();
+        }
+
+        /// <inheritdoc />
+        public ConsensusRules Register<TRule>() where TRule : ConsensusRule, new()
+        {
+            ConsensusRule rule = new TRule()
+            {
+                Parent = this,
+                Logger = this.loggerFactory.CreateLogger(typeof(TRule).FullName)
+            };
+
+            foreach (var dependency in rule.Dependencies())
+            {
+                var depdendency = this.consensusRules.Keys.SingleOrDefault(w => w.GetType() == dependency);
+
+                if (depdendency == null)
+                {
+                    throw new Exception(); // todo
+                }
+            }
+
+            this.consensusRules.Add(rule, rule);
```

Comment:
`rule, rule` makes me thing that this can probably some kind of collection rather than a dictionary.
Or maybe the key should be the name of the rule?

Reply:
yes perhaps `rule.GetType().FullName`